File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_025/java/Solution.java
package com.hi.dhl.algorithms.leetcode._025.java;

import com.hi.dhl.algorithms.model.ListNode;

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/6/21
 *     desc  :
 * </pre>
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        int len = getListLength(head);
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode p = dummy;
        ListNode q = dummy.next;
        int index = 0;
        while (q != null) {
            if (len - index < k) {
                break;
            }
            index = index + k;
            int m = k;
            while (--m != 0) {
                ListNode node = q.next;
                q.next = q.next.next;
                node.next = p.next;
                p.next = node;
            }
            p = q;
            q = q.next;
        }
        return dummy.next;
    }

    private int getListLength(ListNode head) {
        ListNode cur = head;
        int len = 0;
        while (cur != null) {
            len++;
            cur = cur.next;
        }
        return len;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_025/kotlin/Solution.kt
package com.hi.dhl.algorithms.leetcode._025.kotlin

import com.hi.dhl.algorithms.model.ListNode

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/6/24
 *     desc  :
 * </pre>
 */
class Solution {

    fun reverseKGroup(head: ListNode?, k: Int): ListNode? {
        if (head == null) {
            return head
        }
        val len = getListLength(head)
        val dummy = ListNode(0)
        dummy.next = head
        var p = dummy
        var q = dummy.next
        var index = 0
        while (q != null) {
            if (len - index < k) {
                return dummy.next
            }
            index = index + k
            var m = k;
            while (--m != 0) {
                val node = q.next
                q.next = q.next.next
                node.next = p.next
                p.next = node
            }
            p = q
            q = q.next
        }
        return dummy.next
    }

    private fun getListLength(head: ListNode?): Int {
        var cur = head
        var len = 0
        while (cur != null) {
            cur = cur?.next
            len++
        }
        return len
    }
}
